beamer/middleware.py

```diff
@@ -41,3 +51,221 @@ def middleware(method: RPCEndpoint, params: Any) -> RPCResponse:
+# the idea that we ramp up to the original normal operation request rate (i.e.
+# without any delays). Currently,t the ramp up is implemented in the following
+# way:
+#
+# - upon exiting the rate limiting mode, set
+#   - _RateLimiterState.taper_counter to 0
+#   - _RateLimiterState.taper_counter_max to the number of threads currently
+#    waiting on the rate limiter lock in the _rate_limiter function
+#
+# - as each thread that waited on the lock gets its turn, it will notice it is
+#   in the tapering mode (taper_counter_max > 0) so it will sleep for a delay of
+#
+#       _RATE_LIMIT_REQUEST_DELAY / 2**state.taper_counter
+#
+# - when taper_counter reaches taper_counter_max, the tapering mode is over and
+#   we return to normal operation
+#
+# - if at any point during tapering mode we get rate limited by the RPC, we
+#   will immediately switch back to rate limiting mode
+#
+# The tapering mode tries to quickly ramp up to normal request rate by
+# exponentially reducing the delay time between consecutive requests.
+
+
+# The rate limiter state.
+#
+# One instance per web3 rate limiter middleware, i.e. one instance per
+# chain/RPC if there is exactly one Web3 instance per chain/RPC.
+#
+# Note: it is important to ensure only one Web3 instance per chain/RPC,
+# otherwise, it could happen that two parallel rate limiter middleware talk to
+# the same RPC, which would rate limit both middlewares.
+#
+# The lock protects all state fields except num_waiting_on_lock, which is an
+# atomic used to keep track of threads waiting on the lock.
+@dataclass(slots=True)
+class _RateLimiterState:
+    lock: threading.Lock = field(default_factory=threading.Lock)
+    num_waiting_on_lock: atomics.base.AtomicUint = field(
+        default_factory=lambda: atomics.atomic(width=4, atype=atomics.UINT)
+    )
+    rate_limit_end: None | float = None
+    taper_counter: int = 0
+    taper_counter_max: int = 0
+
+
+# The time period during which the middleware will limit the maximum number of
+# requests per second to 1 / _RATE_LIMIT_REQUEST_DELAY.
+# _RATE_LIMIT_PERIOD must always be greater than _RATE_LIMIT_REQUEST_DELAY.
+# It is recommended to use a small multiple of _RATE_LIMIT_REQUEST_DELAY.
+_RATE_LIMIT_PERIOD = 3
+
+# The delay between two consecutive attempts at making a request to the RPC.
+# Only used during the rate limiting period.
+_RATE_LIMIT_REQUEST_DELAY = 1
+
+# A thread local object used to handle reentrancy in web3py middleware.
+_RATE_LIMITER_TLD = threading.local()
+
+# A type alias to improve readability.
+_MakeRequest = Callable[[RPCEndpoint, Any], RPCResponse]
+
+
+def _try_make_request(
+    make_request: _MakeRequest, method: RPCEndpoint, params: Any
+) -> tuple[bool, None | RPCResponse]:
+    try:
+        response = make_request(method, params)
+    except requests.exceptions.HTTPError as exc:
+        if exc.response.status_code == 429:
+            return True, None
+        raise exc
```

Comment:
In which case this will raise ? I understood it is HTTPError but do we have an example for this? Also, if this raises somehow then other parts will break